Fix NPE for REST /appmaster/0?detail=true when no appmasters have launched

Stack trace is
[INFO] [11/27/2014 01:11:27.776] [MasterProxy] Get msg AppMasterDataDetailRequest, forwarding to akka.tcp://master@127.0.0.1:3000/user/master
[ERROR] [11/27/2014 01:11:27.842] [client-akka.actor.default-dispatcher-15] [akka://client/user/rest-services] Error during processing of request HttpRequest(GET,http://localhost:8090/appmaster/0?detail=true,List(Accept-Language: en-US, en;q=0.8, Accept-Encoding: gzip, deflate, sdch, User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/38.0.2125.122 Safari/537.36, Connection: keep-alive, Host: localhost:8090),Empty,HTTP/1.1)
java.lang.NullPointerException
    at org.apache.gearpump.services.AppMasterProtocol$ApplicationFormat$.write(AppMasterProtocol.scala:93)
    at org.apache.gearpump.services.AppMasterProtocol$ApplicationFormat$.write(AppMasterProtocol.scala:87)
    at spray.json.ProductFormats$class.productElement2Field(ProductFormats.scala:45)
    at org.apache.gearpump.services.AppMasterProtocol$.productElement2Field(AppMasterProtocol.scala:28)
    at spray.json.ProductFormatsInstances$$anon$2.write(ProductFormatsInstances.scala:51)
    at spray.json.ProductFormatsInstances$$anon$2.write(ProductFormatsInstances.scala:46)
    at spray.httpx.SprayJsonSupport$$anonfun$sprayJsonMarshaller$1.apply(SprayJsonSupport.scala:43)
    at spray.httpx.SprayJsonSupport$$anonfun$sprayJsonMarshaller$1.apply(SprayJsonSupport.scala:42)
    at spray.httpx.marshalling.Marshaller$MarshallerDelegation$$anonfun$apply$1.apply(Marshaller.scala:58)
    at spray.httpx.marshalling.Marshaller$MarshallerDelegation$$anonfun$apply$1.apply(Marshaller.scala:58)
    at spray.httpx.marshalling.Marshaller$MarshallerDelegation$$anonfun$apply$2.apply(Marshaller.scala:61)
    at spray.httpx.marshalling.Marshaller$MarshallerDelegation$$anonfun$apply$2.apply(Marshaller.scala:60)
    at spray.httpx.marshalling.Marshaller$$anon$2.apply(Marshaller.scala:47)
    at spray.httpx.marshalling.BasicToResponseMarshallers$$anon$1.apply(BasicToResponseMarshallers.scala:35)
    at spray.httpx.marshalling.BasicToResponseMarshallers$$anon$1.apply(BasicToResponseMarshallers.scala:22)
    at spray.httpx.marshalling.ToResponseMarshaller$$anonfun$compose$1.apply(Marshaller.scala:69)
    at spray.httpx.marshalling.ToResponseMarshaller$$anonfun$compose$1.apply(Marshaller.scala:69)
    at spray.httpx.marshalling.ToResponseMarshaller$$anon$3.apply(Marshaller.scala:81)
    at spray.httpx.marshalling.ToResponseMarshallable$$anon$6.marshal(Marshaller.scala:141)
    at spray.httpx.marshalling.ToResponseMarshallable$$anon$7.apply(Marshaller.scala:145)
    at spray.httpx.marshalling.ToResponseMarshallable$$anon$7.apply(Marshaller.scala:144)
    at spray.routing.RequestContext.complete(RequestContext.scala:237)
    at spray.routing.directives.RouteDirectives$$anonfun$complete$1$$anon$3.apply(RouteDirectives.scala:49)
    at spray.routing.directives.RouteDirectives$$anonfun$complete$1$$anon$3.apply(RouteDirectives.scala:48)
    at spray.routing.directives.OnCompleteFutureMagnet$$anon$1$$anonfun$happly$1$$anonfun$apply$1.apply(FutureDirectives.scala:63)
    at spray.routing.directives.OnCompleteFutureMagnet$$anon$1$$anonfun$happly$1$$anonfun$apply$1.apply(FutureDirectives.scala:62)
    at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
    at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:67)
    at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:82)

--- a/rest/src/main/scala/org/apache/gearpump/services/AppMasterProtocol.scala
+++ b/rest/src/main/scala/org/apache/gearpump/services/AppMasterProtocol.scala
@@ -90,7 +90,7 @@ object AppMasterProtocol extends DefaultJsonProtocol  {
         appDescription.toJson
       case _ =>
         JsObject(
-          "appMaster" -> JsString(obj.appMaster),
+          "appMaster" -> JsNull,
           "conf" -> JsNull
         )
     }

--- a/rest/src/test/scala/org/apache/gearpump/services/AppMasterServiceSpec.scala
+++ b/rest/src/test/scala/org/apache/gearpump/services/AppMasterServiceSpec.scala
@@ -19,18 +19,27 @@
 package org.apache.gearpump.services
 
 import org.apache.gearpump.cluster.AppMasterInfo
-import org.apache.gearpump.cluster.MasterToAppMaster.AppMasterData
+import org.apache.gearpump.cluster.MasterToAppMaster.{AppMasterDataDetail, AppMasterData}
+import org.apache.gearpump.services.RestTestUtil.RestTest
+import org.apache.gearpump.streaming.{AppMaster, AppDescription}
+import org.apache.gearpump.util.{Configs, Graph}
+import org.slf4j.{LoggerFactory, Logger}
 import org.specs2.mutable.Specification
-import org.specs2.specification.AfterExample
+import org.specs2.specification.{BeforeExample, AfterExample}
 import spray.testkit.Specs2RouteTest
 
-class AppMasterServiceSpec extends Specification with Specs2RouteTest with AppMasterService with AfterExample {
+class AppMasterServiceSpec extends Specification with Specs2RouteTest
+with AppMasterService with AfterExample with BeforeExample {
   import org.apache.gearpump.services.AppMasterProtocol._
   import spray.httpx.SprayJsonSupport._
   def actorRefFactory = system
   Thread.sleep(500)
-  val restUtil = RestTestUtil.startRestServices
+  val restUtil = before
   val master = restUtil.miniCluster.mockMaster
+
+  def before: RestTest = {
+    RestTestUtil.startRestServices
+  }
 
   "AppMasterService" should {
 //    "return a JSON structure for GET request when detail = true" in {

